Source/console-checkers/Source/ConsoleInputComponent.cpp

```diff
@@ -0,0 +1,229 @@
+		{
+			// If we don't have the right number of arguments we can't even construct this command.
+			if (args.size() != ChangeViewStyleCommand::Settings::s_requiredArgumentCount)
+			{
+				return { nullptr, UIText::s_errorCommandReasonInvalidArument.data()};
+			}
+
+			// We at least have the correct number of arguments, we can try and construct.
+			return { std::make_unique<ChangeViewStyleCommand>(
+				args[ChangeViewStyleCommand::Settings::s_viewStyleArgPosition]) };
+		};
+
+		// Register the move command.
+		m_commandRegistry[s_moveCommandId] =
+			[](const std::vector<std::string>& args) -> CommandResult
+		{
+
+			if (args.size() != MoveCommand::Settings::s_requiredArgumentCount)
+			{
+				return { nullptr, UIText::s_errorCommandReasonInvalidArument.data() };
+			}
+
+			return { std::make_unique<MoveCommand>(
+				args[MoveCommand::Settings::s_sourceArgPosition],
+				args[MoveCommand::Settings::s_destinationArgPosition]) };
+		};
+	}
+
+	~CommandFactory() = default;
+
+	CommandResult CreateCommand(unsigned char cmdId, const std::vector<std::string>& args) const
+	{
+		// C++17 if-init https://en.cppreference.com/w/cpp/language/if
+		if (const auto it = m_commandRegistry.find(cmdId); it != m_commandRegistry.end())
+		{
+			return it->second(args);
+		}
+
+		// Construction failed, errorInfo will be filled.
+		return { nullptr, UIText::s_errorCommandReasonInvalidCommand.data() };
+	}
+
+private:
+
+	// At the point where a command is needed, the appropriate delegate will be called to create the command.
+	using CommandRegistrationDelegate = std::function<CommandResult(const std::vector<std::string>&)>;
+	std::unordered_map<unsigned char, CommandRegistrationDelegate> m_commandRegistry;
+};
+
+ConsoleInputComponent::ConsoleInputComponent(Game* game)
+	: m_game(game)
+	, m_commandFactory(std::make_unique<CommandFactory>())
+{
+}
+ConsoleInputComponent::~ConsoleInputComponent() = default;
+
+//---------------------------------------------------------------
+
+void ConsoleInputComponent::RequestAndProcessInput() const
+{
+	std::string input;
+	std::getline(std::cin, input);
+
+	if (input.empty())
+	{
+		return;
+	}
+
+	std::string sanitizedInput;
+	auto toLower = [](const unsigned char c) { return std::tolower(c); };
+	std::ranges::copy(std::views::transform(input, toLower), std::back_inserter(sanitizedInput));
+
+	// The first character in the input is our identifier for which command to execute.
+	const unsigned char commandId = sanitizedInput[0];
+
+	// Skip over the commandId we just looked at.
+	std::istringstream iss(sanitizedInput.substr(1));
+	std::vector<std::string> args;
+
+	std::string arg;
+	while (iss >> arg)
+	{
+		args.push_back(arg);
+	}
+
+	const auto commandResult = m_commandFactory->CreateCommand(commandId, args);
+	if (!commandResult.command)
+	{
+		spdlog::info("Failed to create command, it will not be executed. commandId={} input={}", commandId, sanitizedInput);
```

Comment:
I chose info because this is not a system error that we care about, this is a user malformed their command input. I tried to use errors for things that shouldn't happen but I'll admit its a bit arbitrary haha